jre_emul/android/platform/libcore/luni/src/test/java/libcore/java/util/jar/JarInputStreamTest.java

```diff
@@ -0,0 +1,441 @@
+        Set<String> actual = new HashSet<String>();
+        InputStream is = new URL(jarName).openConnection().getInputStream();
+        JarInputStream jis = new JarInputStream(is);
+        JarEntry je = jis.getNextJarEntry();
+        while (je != null) {
+            actual.add(je.toString());
+            je = jis.getNextJarEntry();
+        }
+        assertEquals(actual, desired);
+        jis.close();
+
+        try {
+            jis.getNextJarEntry();
+            fail("IOException expected");
+        } catch (IOException ee) {
+            // expected
+        }
+
+        File resources = Support_Resources.createTempFolder();
+        Support_Resources.copyFile(resources, null, "Broken_entry.jar");
+        is = Support_Resources.getStream("Broken_entry.jar");
+        jis = new JarInputStream(is, false);
+        jis.getNextJarEntry();
+        try {
+            jis.getNextJarEntry();
+            fail("ZipException expected");
+        } catch (ZipException ee) {
+            // expected
+        }
+    }
+
+    public void test_getManifest() throws Exception {
+        // Test for method java.util.jar.Manifest
+        // java.util.jar.JarInputStream.getManifest()
+        Manifest m;
+        InputStream is = new URL(jarName2).openConnection().getInputStream();
+        JarInputStream jis = new JarInputStream(is);
+        m = jis.getManifest();
+        assertNull("The jar input stream should not have a manifest", m);
+
+        is = new URL(jarName).openConnection().getInputStream();
+        jis = new JarInputStream(is);
+        m = jis.getManifest();
+        assertNotNull("The jar input stream should have a manifest", m);
+    }
+
+    public void test_getNextJarEntry() throws Exception {
+        final Set<String> desired = new HashSet<String>(Arrays.asList(new String[] { "foo/",
+                "foo/bar/", A_CLASS, "Blah.txt" }));
+        Set<String> actual = new HashSet<String>();
+        InputStream is = new URL(jarName).openConnection().getInputStream();
+        JarInputStream jis = new JarInputStream(is);
+        JarEntry je = jis.getNextJarEntry();
+        while (je != null) {
+            actual.add(je.toString());
+            je = jis.getNextJarEntry();
+        }
+        assertEquals(actual, desired);
+    }
+
+    public void test_JarInputStream_Integrate_Jar_getNextEntry()
+            throws IOException {
+        String intJarName = Support_Resources.getURL("Integrate.jar");
+        InputStream is = new URL(intJarName).openConnection()
+                .getInputStream();
+        JarInputStream jin = new JarInputStream(is, true);
+        ZipEntry entry = null;
+        int count = 0;
+        while (count == 0 || entry != null) {
+            count++;
+            entry = jin.getNextEntry();
+        }
+        assertEquals(TOTAL_ENTRIES + 1, count);
+        jin.close();
+    }
+
+    /**
+     * These test the modified jar feature j2objc doesn't support,
```

Comment:
Add "J2ObjC removed" for easier future upgrade work?

Reply:
added the comment